File: test/org/newtonpolyhedron/solve/cone/ConeSolverImplTest.java
package org.newtonpolyhedron.solve.cone;

import static org.fs.test.Assert.*;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.junit.Test;
import org.newtonpolyhedron.entity.vector.IntVector;
import org.newtonpolyhedron.solve.cone.ConeSolver;
import org.newtonpolyhedron.solve.cone.ConeSolverImpl;
import org.newtonpolyhedron.utils.NullPrintWriter;

public class ConeSolverImplTest {
	
	private static final ConeSolver	coneSolver	= new ConeSolverImpl();
	
	
	
	@Test
	public void chernTestCase() throws Exception {
		List <IntVector> eqSys = new ArrayList <IntVector>();
		eqSys.add(new IntVector(1, -1, 3, -8));
		eqSys.add(new IntVector(-1, 2, -1, 1));
		eqSys.add(new IntVector(2, -1, -2, 1));
		eqSys.add(new IntVector(-3, 1, -1, 6));
		eqSys.add(new IntVector(1, 1, -3, 2));
		List <IntVector> solution = coneSolver.solve(eqSys, null, 4, NullPrintWriter.instance);
		assertSetEquals(//
				Arrays.asList(//
						new IntVector(15, 11, 12, 5),//
						new IntVector(13, 9, 12, 7),//
						new IntVector(5, 5, 8, 3),//
						new IntVector(1, -1, 2, 1) //
				), solution);
	}
	
	@Test
	public void chernTestCaseMod() throws Exception {
		List <IntVector> eqSys = new ArrayList <IntVector>();
		eqSys.add(new IntVector(1, -1, 3, -8, 0));
		eqSys.add(new IntVector(-1, 2, -1, 1, 0));
		eqSys.add(new IntVector(2, -1, -2, 1, 0));
		eqSys.add(new IntVector(-3, 1, -1, 6, 0));
		eqSys.add(new IntVector(1, 1, -3, 2, 0));
		List <IntVector> solution = coneSolver.solve(eqSys, null, 5, NullPrintWriter.instance);
		assertSetEquals(//
				Arrays.asList(//
						new IntVector(15, 11, 12, 5, 0),//
						new IntVector(13, 9, 12, 7, 0),//
						new IntVector(5, 5, 8, 3, 0),//
						new IntVector(1, -1, 2, 1, 0) //
				), solution);
	}
	
	@Test
	public void chernTestCase5dOctant() throws Exception {
		List <IntVector> eqSys = new ArrayList <IntVector>();
		eqSys.add(new IntVector(-1, 0, 0, 0, 0));
		eqSys.add(new IntVector(0, -1, 0, 0, 0));
		eqSys.add(new IntVector(0, 0, -1, 0, 0));
		eqSys.add(new IntVector(0, 0, 0, -1, 0));
		eqSys.add(new IntVector(0, 0, 0, 0, -1));
		eqSys.add(new IntVector(1, -1, 3, -8, 5));
		eqSys.add(new IntVector(-1, 2, -1, 1, -1));
		eqSys.add(new IntVector(2, -1, -2, 1, 0));
		eqSys.add(new IntVector(-3, 1, -1, 6, -3));
		eqSys.add(new IntVector(1, 1, -3, 2, -1));
		List <IntVector> solution = coneSolver.solve(eqSys, null, 5, NullPrintWriter.instance);
		assertSetEquals(//
				Arrays.asList(//
						new IntVector(15, 11, 12, 5, 0),//
						new IntVector(13, 0, 17, 8, 0),//
						new IntVector(5, 5, 8, 3, 0),//
						new IntVector(1, 1, 1, 1, 1),//
						new IntVector(13, 9, 12, 7, 0),//
						new IntVector(11, 0, 15, 8, 0),//
						new IntVector(2, 0, 3, 2, 1),//
						new IntVector(5, 0, 9, 4, 0) //
				), solution);
	}
	
	
	
	@Test
	public void simpleBrunoTestCase() throws Exception {
		List <IntVector> eqSys = new ArrayList <IntVector>();
		eqSys.add(new IntVector(3, -1, -1));
		eqSys.add(new IntVector(-1, 3, -1));
		eqSys.add(new IntVector(-1, -1, 3));
		eqSys.add(new IntVector(1, -1, 1));
		List <IntVector> solution = coneSolver.solve(eqSys, null, 3, NullPrintWriter.instance);
		assertSetEquals(//
				Arrays.asList(//
						new IntVector(-1, -2, -1),//
						new IntVector(-1, -1, -2),//
						new IntVector(-2, -1, -1) //
				), solution);
	}
	
	@Test
	public void simpleBrunoTestCase2() throws Exception {
		List <IntVector> eqSys = new ArrayList <IntVector>();
		eqSys.add(new IntVector(3, -1, -1));
		eqSys.add(new IntVector(-1, 3, -1));
		eqSys.add(new IntVector(3, -1, 2));
		eqSys.add(new IntVector(-1, 3, 2));
		List <IntVector> solution = coneSolver.solve(eqSys, null, 3, NullPrintWriter.instance);
		assertSetEquals(//
				Arrays.asList(//
						new IntVector(-1, -1, -2),//
						new IntVector(-1, -3, 0),//
						new IntVector(-3, -1, 0),//
						new IntVector(-1, -1, 1) //
				), solution);
	}
	
	
	
	@Test
	public void brunoTestCase() throws Exception {
		List <IntVector> eqSys = new ArrayList <IntVector>();
		eqSys.add(new IntVector(-5, -1, -1, -1));
		eqSys.add(new IntVector(-4, -2, -1, -1));
		eqSys.add(new IntVector(-4, 0, -2, -2));
		eqSys.add(new IntVector(-4, 0, 0, 0));
		eqSys.add(new IntVector(-4, 4, 0, 0));
		eqSys.add(new IntVector(-4, 0, 4, 0));
		eqSys.add(new IntVector(-4, 0, 0, 4));
		eqSys.add(new IntVector(-8, 0, 0, 0));
		eqSys.add(new IntVector(-4, -4, 0, 0));
		eqSys.add(new IntVector(-4, 0, -4, 0));
		eqSys.add(new IntVector(-4, 0, 0, -4));
		
		List <IntVector> solution = coneSolver.solve(eqSys, null, 4, NullPrintWriter.instance);
		assertSetEquals(//
				Arrays.asList(//
						new IntVector(1, -1, -1, -1),//
						new IntVector(1, 1, -1, -1),//
						new IntVector(1, 1, 1, 1),//
						new IntVector(1, -1, 1, 1),//
						new IntVector(1, 1, -1, 1),// A
						new IntVector(1, -1, -1, 1),//
						new IntVector(1, 1, 1, -1),//
						new IntVector(1, -1, 1, -1) //
				), solution);
	}
	
	@Test
	public void brunoTestCase2() throws Exception {
		List <IntVector> eqSys = new ArrayList <IntVector>();
		eqSys.add(new IntVector(2, 0, 0, 0));
		eqSys.add(new IntVector(2, 1, 0, 0));
		eqSys.add(new IntVector(2, 1, 1, 1));
		eqSys.add(new IntVector(2, 2, 1, 1));
		eqSys.add(new IntVector(2, 1, 2, 1));
		
		List <IntVector> solution = coneSolver.solve(eqSys, null, 4, NullPrintWriter.instance);
		assertSetEquals(//
				Arrays.asList(//
						new IntVector(0, 0, -1, 1),//
						new IntVector(0, 0, 1, -2),//
						new IntVector(0, -1, 0, 1),//
						new IntVector(-1, 2, 2, -4),//
						new IntVector(-1, 0, 0, 2) //
				), solution);
	}
	
	
	@Test
	public void simple2dTestCase1() throws Exception {
		List <IntVector> eqSys = new ArrayList <IntVector>();
		eqSys.add(new IntVector(1, -4));
		eqSys.add(new IntVector(-2, -1));
		List <IntVector> solution = coneSolver.solve(eqSys, null, 2, NullPrintWriter.instance);
		assertSetEquals(//
				Arrays.asList(//
						new IntVector(4, 1),//
						new IntVector(-1, 2) //
				), solution);
	}
	
	@Test
	public void simple2dTestCase2() throws Exception {
		List <IntVector> eqSys = new ArrayList <IntVector>();
		eqSys.add(new IntVector(-1, 2));
		eqSys.add(new IntVector(2, -1));
		List <IntVector> solution = coneSolver.solve(eqSys, null, 2, NullPrintWriter.instance);
		assertSetEquals(//
				Arrays.asList(//
						new IntVector(-2, -1),//
						new IntVector(-1, -2) //
				), solution);
	}
	
	@Test
	public void simple2dTestCase3() throws Exception {
		List <IntVector> eqSys = new ArrayList <IntVector>();
		eqSys.add(new IntVector(-1, 2));
		eqSys.add(new IntVector(3, -3));
		List <IntVector> solution = coneSolver.solve(eqSys, null, 2, NullPrintWriter.instance);
		assertSetEquals(//
				Arrays.asList(//
						new IntVector(-2, -1),//
						new IntVector(-1, -1) //
				), solution);
	}
	
	
	
	@Test
	public void simple3dTestCase1() throws Exception {
		List <IntVector> eqSys = new ArrayList <IntVector>();
		eqSys.add(new IntVector(-3, 1, 1));
		eqSys.add(new IntVector(-1, 1, 0));
		eqSys.add(new IntVector(-1, 0, 1));
		List <IntVector> solution = coneSolver.solve(eqSys, null, 3, NullPrintWriter.instance);
		assertSetEquals(//
				Arrays.asList(//
						new IntVector(-1, -1, -2),//
						new IntVector(-1, -2, -1),//
						new IntVector(1, 1, 1) //
				), solution);
	}
	
	@Test
	public void simple3dTestCase2() throws Exception {
		List <IntVector> eqSys = new ArrayList <IntVector>();
		eqSys.add(new IntVector(-3, 1, 1));
		eqSys.add(new IntVector(-1, 1, 0));
		eqSys.add(new IntVector(-1, 0, 1));
		List <IntVector> solution = coneSolver.solve(eqSys, null, 3, NullPrintWriter.instance);
		assertSetEquals(//
				Arrays.asList(//
						new IntVector(-1, -1, -2),//
						new IntVector(-1, -2, -1),//
						new IntVector(1, 1, 1) //
				), solution);
	}
	
	
	
	@Test
	public void randomNullBasisCase1() throws Exception {
		List <IntVector> eqSys = new ArrayList <IntVector>();
		eqSys.add(new IntVector(-9, 8, 0));
		eqSys.add(new IntVector(-9, 0, 7));
		eqSys.add(new IntVector(-6, 2, 1));
		eqSys.add(new IntVector(-3, 4, 0));
		eqSys.add(new IntVector(-3, 0, 5));
		eqSys.add(new IntVector(-2, 2, 2));
		List <IntVector> solution = coneSolver.solve(eqSys, null, 3, NullPrintWriter.instance);
		assertSetEquals(//
				Arrays.asList(//
						new IntVector(-8, -9, -30),//
						new IntVector(-14, -33, -18),//
						new IntVector(0, 0, -1),//
						new IntVector(0, -1, 0),//
						new IntVector(4, 3, 1),//
						new IntVector(5, 2, 3) //
				), solution);
	}
	
	@Test
	public void randomNullBasisCase2() throws Exception {
		List <IntVector> eqSys = new ArrayList <IntVector>();
		eqSys.add(new IntVector(9, -8, 0));
		eqSys.add(new IntVector(0, -8, 7));
		eqSys.add(new IntVector(3, -6, 1));
		eqSys.add(new IntVector(3, -4, 0));
		eqSys.add(new IntVector(0, -4, 5));
		eqSys.add(new IntVector(1, -2, 2));
		List <IntVector> solution = coneSolver.solve(eqSys, null, 3, NullPrintWriter.instance);
		assertSetEquals(//
				Arrays.asList(//
						new IntVector(-34, -21, -24),//
						new IntVector(0, 0, -1),//
						new IntVector(-4, -3, -6),//
						new IntVector(-1, 0, 0),//
						new IntVector(8, 9, 5),//
						new IntVector(2, 5, 4) //
				), solution);
	}
	
	@Test
	public void randomNullBasisCase3() throws Exception {
		List <IntVector> eqSys = new ArrayList <IntVector>();
		eqSys.add(new IntVector(9, 0, -7));
		eqSys.add(new IntVector(0, 8, -7));
		eqSys.add(new IntVector(3, 2, -6));
		eqSys.add(new IntVector(3, 0, -5));
		eqSys.add(new IntVector(0, 4, -5));
		eqSys.add(new IntVector(1, 2, -3));
		List <IntVector> solution = coneSolver.solve(eqSys, null, 3, NullPrintWriter.instance);
		assertSetEquals(//
				Arrays.asList(//
						new IntVector(56, 63, 72),//
						new IntVector(0, -1, 0),//
						new IntVector(-1, 0, 0),//
						new IntVector(-20, -15, -12) //
				), solution);
	}
}

File: test/org/newtonpolyhedron/solve/cone/ConeSolverImplTest.scala
package org.newtonpolyhedron.solve.cone

import org.junit.runner.RunWith
import org.scalatest.FunSuite
import org.scalatest.junit.JUnitRunner
import org.newtonpolyhedron.entity.vector.IntMathVec
import org.newtonpolyhedron.utils.NullPrintWriter

@RunWith(classOf[JUnitRunner])
class ConeSolverImplTest extends FunSuite {

  val coneSolver = new ConeSolverImpl

  implicit def tupleToVec(tuple: Tuple2[Int, Int]) =
    IntMathVec(tuple._1, tuple._2)

  implicit def tupleToVec(tuple: Tuple3[Int, Int, Int]) =
    IntMathVec(tuple._1, tuple._2, tuple._3)

  implicit def tupleToVec(tuple: Tuple4[Int, Int, Int, Int]) =
    IntMathVec(tuple._1, tuple._2, tuple._3, tuple._4)

  implicit def tupleToVec(tuple: Tuple5[Int, Int, Int, Int, Int]) =
    IntMathVec(tuple._1, tuple._2, tuple._3, tuple._4, tuple._5)

  test("chernyakov") {
    def eqSys = IndexedSeq(
      (1, -1, 3, -8),
      (-1, 2, -1, 1),
      (2, -1, -2, 1),
      (-3, 1, -1, 6),
      (1, 1, -3, 2)) map tupleToVec
    val solution = coneSolver.solve(eqSys, IndexedSeq.empty, 4, NullPrintWriter.instance);
    assert(solution.toSet === (IndexedSeq(
      (15, 11, 12, 5),
      (13, 9, 12, 7),
      (5, 5, 8, 3),
      (1, -1, 2, 1)) map tupleToVec).toSet)
  }

  test("chernyakov degenerated 4d") {
    def eqSys = IndexedSeq(
      (1, -1, 3, 0),
      (-1, 2, -1, 0),
      (2, -1, -2, 0),
      (-3, 1, -1, 0),
      (1, 1, -3, 0)) map tupleToVec
    val solution = coneSolver.solve(eqSys, IndexedSeq.empty, 4, NullPrintWriter.instance);
    assert(solution.toSet === (IndexedSeq(
      (0, 0, 0, 1),
      (0, 0, 0, -1)) map tupleToVec).toSet)
  }

  test("chernyakov degenerated 5d") {
    def eqSys = IndexedSeq(
      (1, -1, 3, -8, 0),
      (-1, 2, -1, 1, 0),
      (2, -1, -2, 1, 0),
      (-3, 1, -1, 6, 0),
      (1, 1, -3, 2, 0)) map tupleToVec
    val solution = coneSolver.solve(eqSys, IndexedSeq.empty, 5, NullPrintWriter.instance);
    assert(solution.toSet === (IndexedSeq(
      (0, 0, 0, 0, 1),
      (0, 0, 0, 0, -1)) map tupleToVec).toSet)
  }

  test("chernyakov 5d octant") {
    def eqSys = IndexedSeq(
      (-1, 0, 0, 0, 0),
      (0, -1, 0, 0, 0),
      (0, 0, -1, 0, 0),
      (0, 0, 0, -1, 0),
      (0, 0, 0, 0, -1),
      (1, -1, 3, -8, 5),
      (-1, 2, -1, 1, -1),
      (2, -1, -2, 1, 0),
      (-3, 1, -1, 6, -3),
      (1, 1, -3, 2, -1)) map tupleToVec
    val solution = coneSolver.solve(eqSys, IndexedSeq.empty, 5, NullPrintWriter.instance);
    assert(solution.toSet === (IndexedSeq(
      (15, 11, 12, 5, 0),
      (13, 0, 17, 8, 0),
      (5, 5, 8, 3, 0),
      (1, 1, 1, 1, 1),
      (13, 9, 12, 7, 0),
      (11, 0, 15, 8, 0),
      (2, 0, 3, 2, 1),
      (5, 0, 9, 4, 0)) map tupleToVec).toSet)
  }

  //
  //
  //
  test("bruno simple test case") {
    def eqSys = IndexedSeq(
      (3, -1, -1),
      (-1, 3, -1),
      (-1, -1, 3),
      (1, -1, 1)) map tupleToVec
    val solution = coneSolver.solve(eqSys, IndexedSeq.empty, 3, NullPrintWriter.instance);
    assert(solution.toSet === (IndexedSeq(
      (-1, -2, -1),
      (-1, -1, -2),
      (-2, -1, -1)) map tupleToVec).toSet)
  }

  test("bruno simple test case 2") {
    def eqSys = IndexedSeq(
      (3, -1, -1),
      (-1, 3, -1),
      (3, -1, 2),
      (-1, 3, 2)) map tupleToVec
    val solution = coneSolver.solve(eqSys, IndexedSeq.empty, 3, NullPrintWriter.instance);
    assert(solution.toSet === (IndexedSeq(
      (-1, -1, -2),
      (-1, -3, 0),
      (-3, -1, 0),
      (-1, -1, 1)) map tupleToVec).toSet)
  }

  //
  //
  //
  test("bruno test case") {
    def eqSys = IndexedSeq(
      (-5, -1, -1, -1),
      (-4, -2, -1, -1),
      (-4, 0, -2, -2),
      (-4, 0, 0, 0),
      (-4, 4, 0, 0),
      (-4, 0, 4, 0),
      (-4, 0, 0, 4),
      (-8, 0, 0, 0),
      (-4, -4, 0, 0),
      (-4, 0, -4, 0),
      (-4, 0, 0, -4)) map tupleToVec
    val solution = coneSolver.solve(eqSys, IndexedSeq.empty, 4, NullPrintWriter.instance);
    assert(solution.toSet === (IndexedSeq(
      (1, -1, -1, -1),
      (1, 1, -1, -1),
      (1, 1, 1, 1),
      (1, -1, 1, 1),
      (1, 1, -1, 1),
      (1, -1, -1, 1),
      (1, 1, 1, -1),
      (1, -1, 1, -1)) map tupleToVec).toSet)
  }

  test("bruno test case 2") {
    def eqSys = IndexedSeq(
      (2, 0, 0, 0),
      (2, 1, 0, 0),
      (2, 1, 1, 1),
      (2, 2, 1, 1),
      (2, 1, 2, 1)) map tupleToVec
    val solution = coneSolver.solve(eqSys, IndexedSeq.empty, 4, NullPrintWriter.instance);
    assert(solution.toSet === (IndexedSeq(
      (0, 0, -1, 1),
      (0, 0, 1, -2),
      (0, -1, 0, 1),
      (-1, 2, 2, -4),
      (-1, 0, 0, 2)) map tupleToVec).toSet)
  }

  //
  //
  //
  test("simple 2d") {
    def eqSys = IndexedSeq(
      (1, -4),
      (-2, -1)) map tupleToVec
    val solution = coneSolver.solve(eqSys, IndexedSeq.empty, 2, NullPrintWriter.instance);
    assert(solution.toSet === (IndexedSeq(
      (4, 1),
      (-1, 2)) map tupleToVec).toSet)
  }

  test("simple 2d - 2") {
    def eqSys = IndexedSeq(
      (-1, 2),
      (2, -1)) map tupleToVec
    val solution = coneSolver.solve(eqSys, IndexedSeq.empty, 2, NullPrintWriter.instance);
    assert(solution.toSet === (IndexedSeq(
      (-2, -1),
      (-1, -2)) map tupleToVec).toSet)
  }

  test("simple 2d - 3") {
    def eqSys = IndexedSeq(
      (-1, 2),
      (3, -3)) map tupleToVec
    val solution = coneSolver.solve(eqSys, IndexedSeq.empty, 2, NullPrintWriter.instance);
    assert(solution.toSet === (IndexedSeq(
      (-2, -1),
      (-1, -1)) map tupleToVec).toSet)
  }

  //
  //
  //
  test("simple 3d") {
    def eqSys = IndexedSeq(
      (-3, 1, 1),
      (-1, 1, 0),
      (-1, 0, 1)) map tupleToVec
    val solution = coneSolver.solve(eqSys, IndexedSeq.empty, 3, NullPrintWriter.instance);
    assert(solution.toSet === (IndexedSeq(
      (-1, -1, -2),
      (-1, -2, -1),
      (1, 1, 1)) map tupleToVec).toSet)
  }

  //
  //
  //
  test("random case") {
    def eqSys = IndexedSeq(
      (-9, 8, 0),
      (-9, 0, 7),
      (-6, 2, 1),
      (-3, 4, 0),
      (-3, 0, 5),
      (-2, 2, 2)) map tupleToVec
    val solution = coneSolver.solve(eqSys, IndexedSeq.empty, 3, NullPrintWriter.instance);
    assert(solution.toSet === (IndexedSeq(
      (-8, -9, -30),
      (-14, -33, -18),
      (0, 0, -1),
      (0, -1, 0),
      (4, 3, 1),
      (5, 2, 3)) map tupleToVec).toSet)
  }

  test("random case 2") {
    def eqSys = IndexedSeq(
      (9, -8, 0),
      (0, -8, 7),
      (3, -6, 1),
      (3, -4, 0),
      (0, -4, 5),
      (1, -2, 2)) map tupleToVec
    val solution = coneSolver.solve(eqSys, IndexedSeq.empty, 3, NullPrintWriter.instance);
    assert(solution.toSet === (IndexedSeq(
      (-34, -21, -24),
      (0, 0, -1),
      (-4, -3, -6),
      (-1, 0, 0),
      (8, 9, 5),
      (2, 5, 4)) map tupleToVec).toSet)
  }

  test("random case 3") {
    def eqSys = IndexedSeq(
      (9, 0, -7),
      (0, 8, -7),
      (3, 2, -6),
      (3, 0, -5),
      (0, 4, -5),
      (1, 2, -3)) map tupleToVec
    val solution = coneSolver.solve(eqSys, IndexedSeq.empty, 3, NullPrintWriter.instance);
    assert(solution.toSet === (IndexedSeq(
      (56, 63, 72),
      (0, -1, 0),
      (-1, 0, 0),
      (-20, -15, -12)) map tupleToVec).toSet)
  }
}